### ISSUE TITLE
softmax(input, dim) doesn't need to save `input` for backward computation

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#49600]
## 🐛 Bug

Right now, softmax (and logsoftmax) save `input` for backward computation:
```cpp
//torch/csrc/autograd/generated/VariableTypeEverything.cpp
Tensor _log_softmax(const Tensor & self, int64_t dim, bool half_to_float) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  (void)_any_requires_grad;
  std::shared_ptr<LogSoftmaxBackward> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<LogSoftmaxBackward>(new LogSoftmaxBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
    grad_fn->dim = dim;
  }
  ...
}
```
However, the backward implementation for softmax (softmax_backward_cuda and sofmax_backward_cpu) don't use their `input` argument at all, instead relying on the `output` argument!
https://github.com/pytorch/pytorch/blob/72b00a8a52b570f4248eae26a46873b8f5277000/aten/src/ATen/native/SoftMax.cpp#L168-L198
https://github.com/pytorch/pytorch/blob/72b00a8a52b570f4248eae26a46873b8f5277000/aten/src/ATen/native/cuda/SoftMax.cu#L880-L888

However, we can't excise the argument because it is used by our codegen to compute second order gradients for softmax: 
https://github.com/pytorch/pytorch/blob/72b00a8a52b570f4248eae26a46873b8f5277000/tools/autograd/derivatives.yaml#L1651-L1653

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

**Comment by ak-source** [comment]
same story with softplus

**Comment by vadimkantorov** [comment]
could it have an argument to omit saving it (if the user does not intend any double backward)?

**Comment by cpuhrsch** [comment]
cc @ngimel @fmassa - is this relevant to ongoing discussions about softmax memory footprint?

**Comment by ngimel** [comment]
Yes, #64000 is a dupe of this bug.
